src/ray/raylet/worker_pool.cc

```diff
@@ -587,68 +577,82 @@ void WorkerPool::PushWorker(const std::shared_ptr<WorkerInterface> &worker) {
       state.idle.insert(worker);
+      if (RayConfig::instance().enable_multi_tenancy()) {
+        idle_of_all_languages.push_back(worker);
+      }
     } else {
       state.idle_actor[worker->GetActorId()] = worker;
     }
   }
 }
 
-void WorkerPool::TryKillingIdleWorker(std::shared_ptr<WorkerInterface> worker) {
-  auto &worker_state = GetStateForLanguage(worker->GetLanguage());
-  if (worker_state.pending_unregistration_workers.count(worker) > 0) {
-    // This worker has already been killed.
-    // This is possible because a Java worker process may hold multiple workers.
-    return;
-  }
-
+void WorkerPool::TryKillingIdleWorkers() {
   auto running_size = GetAllRegisteredWorkers().size();
-  for (const auto &entry : states_by_lang_) {
-    running_size -= entry.second.pending_unregistration_workers.size();
-  }
-  if (running_size <= static_cast<size_t>(num_workers_soft_limit_)) {
-    return;
-  }
+  running_size -= pending_unregistration_workers.size();
+
+  // Kill idle workers in FIFO order.
+  for (auto it = idle_of_all_languages.begin();
+       it != idle_of_all_languages.end() &&
+       running_size > static_cast<size_t>(num_workers_soft_limit_);
+       it++) {
+    auto process = (*it)->GetProcess();
+    if (pending_unregistration_workers.count(*it) > 0) {
+      // This worker has already been killed.
+      // This is possible because a Java worker process may hold multiple workers.
+      continue;
+    }
 
-  auto worker_id = worker->WorkerId();
-  const auto pid = worker->GetProcess().GetId();
-  if (worker_state.idle.count(worker) == 0) {
-    return;
-  }
-  if (worker_state.starting_worker_processes.count(worker->GetProcess()) > 0) {
-    // A Java worker process may hold multiple workers.
-    RAY_LOG(DEBUG) << "Some workers of pid " << pid
-                   << " are pending registration. Skip killing worker " << worker_id;
-    return;
-  }
+    auto &worker_state = GetStateForLanguage((*it)->GetLanguage());
 
-  // Make sure all workers in this worker process are idle.
-  // This block of code is needed by Java workers.
-  std::unordered_set<std::shared_ptr<WorkerInterface>> workers_in_the_same_process;
-  for (const auto &worker_in_the_same_process : worker_state.registered_workers) {
-    if (worker_in_the_same_process->GetProcess().GetId() == pid) {
-      if (worker_state.idle.count(worker_in_the_same_process) == 0) {
-        // Another worker in this process isn't idle, so this process can't be killed.
-        return;
-      } else {
+    if (worker_state.starting_worker_processes.count(process) > 0) {
+      // A Java worker process may hold multiple workers.
+      // Some workers of this process are pending registration. Skip killing this worker.
+      continue;
+    }
+
+    // Make sure all workers in this worker process are idle.
+    // This block of code is needed by Java workers.
+    std::unordered_set<std::shared_ptr<WorkerInterface>> workers_in_the_same_process;
```

Comment:
It would be great to move this into a helper function if possible.

Reply:
done